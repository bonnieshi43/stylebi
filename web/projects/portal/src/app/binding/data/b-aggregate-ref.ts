/*
 * inetsoft-web - StyleBI is a business intelligence web application.
 * Copyright Â© 2024 InetSoft Technology (info@inetsoft.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import { AbstractBindingRef } from "./abstract-binding-ref";
import { DataRef } from "../../common/data/data-ref";
import {FormulaOptionModel} from "./formula-option-Model";
import { FormulaInfo } from "./formula-info";
import { CalculateInfo } from "./chart/calculate-info";

export class BAggregateRef extends AbstractBindingRef {
   refType?: number;
   formula: string;
   percentage: string;
   num: number;
   numValue: string;
   secondaryColumnValue: string;
   secondaryColumn: DataRef;
   columnValue: string;
   caption: string;
   calculateInfo?: CalculateInfo;
   formulaOptionModel: FormulaOptionModel;
   originalDataType?: string;
}
