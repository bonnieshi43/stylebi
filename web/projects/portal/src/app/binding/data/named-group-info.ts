/*
 * inetsoft-web - StyleBI is a business intelligence web application.
 * Copyright Â© 2024 InetSoft Technology (info@inetsoft.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affrero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import { ConditionExpression } from "../../common/data/condition/condition-expression";

export class NamedGroupInfo {
   type: number;
   name: string;
   groups: GroupCondition[];
   conditions: ConditionExpression[];

   static EXPERT_NAMEDGROUP_INFO: number = 1;
   static SIMPLE_NAMEDGROUP_INFO: number = 2;
   static ASSET_NAMEDGROUP_INFO: number = 3;
   static ASSET_NAMEDGROUP_INFO_REF: number = 4;
}

export class GroupCondition {
   name: string;
   value: Object[];
}
