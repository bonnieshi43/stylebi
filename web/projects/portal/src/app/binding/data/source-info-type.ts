/*
 * inetsoft-web - StyleBI is a business intelligence web application.
 * Copyright Â© 2024 InetSoft Technology (info@inetsoft.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affrero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
export const enum SourceInfoType {
   /**
    * A normal query source.
    */
   QUERY = 0,
   /**
    * A data model source.
    */
   MODEL = 1,
   /**
    * A source from other bindable report elements in the same report.
    */
   REPORT = 2,
   /**
    * A normal source from report parameters.
    */
   PARAMETER = 3,
   /**
    * A local query source.
    */
   LOCAL_QUERY = 4,
   /**
    * A cube source.
    */
   CUBE = 5,
   /**
    * An embedded data source.
    */
   EMBEDDED_DATA = 6,
   /**
    * An asset source.
    */
   ASSET = 7,
   /**
    * A physical table source.
    */
   PHYSICAL_TABLE = 8,
   /**
    * A source from other bindable assembly elements in the same viewsheet.
    */
   VS_ASSEMBLY = 64,
   /**
    * NULL source.
    */
   NONE = -1
}
