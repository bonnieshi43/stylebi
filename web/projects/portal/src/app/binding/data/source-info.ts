/*
 * This file is part of StyleBI.
 * Copyright (C) 2024  InetSoft Technology
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
export class SourceInfo {
   source: string;
   prefix: string;
   type: number;
   view?: string;
   dataSourceType: string;
   sqlServer?: boolean;
   supportFullOutJoin: boolean;
   rest: boolean;
   browsable?: boolean;
   joinSources: Array<SourceInfo>;

   static NONE = -1;
   static QUERY = 0;
   static MODEL = 1;
   static REPORT = 2;
   static PARAMETER = 3;
   static LOCAL_QUERY = 4;
   static EMBEDDED_DATA = 6;
   static ASSET = 7;
   static PHYSICAL_TABLE = 8;

   static EMBEDED_DATA_NAME = "EMBEDED_DATA";
}
