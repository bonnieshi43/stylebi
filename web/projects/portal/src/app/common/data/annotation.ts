/*
 * inetsoft-web - StyleBI is a business intelligence web application.
 * Copyright Â© 2024 InetSoft Technology (info@inetsoft.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
/**
 * Types of annotation corresponding to AnnotationVSAssemblyInfo.java
 */
export namespace Annotation {
   export type Type = AnnotationType;
   enum AnnotationType {
      VIEWSHEET = 0,
      ASSEMBLY = 1,
      DATA = 2
   }

   export function isViewsheetAnnotation(type: AnnotationType): boolean {
      return type === AnnotationType.VIEWSHEET;
   }

   export function isAssemblyAnnotation(type: AnnotationType): boolean {
      return type === AnnotationType.ASSEMBLY;
   }

   export function isDataAnnotation(type: AnnotationType): boolean {
      return type === AnnotationType.DATA;
   }
}
