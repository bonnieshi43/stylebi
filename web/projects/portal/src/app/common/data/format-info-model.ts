/*
 * This file is part of StyleBI.
 * Copyright (C) 2024  InetSoft Technology
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
import { Format } from "../util/format";

export interface FormatInfoModel {
   type: string; // For JsonTypeInfo, sub-interface should override
   color: string;
   backgroundColor: string;
   font: FontInfo;
   align: AlignmentInfo;
   format: Format | string;
   formatSpec?: string;
   dateSpec?: string;
   borderColor?: string;
   borderTopStyle?: string;
   borderTopColor?: string;
   borderTopWidth?: string;
   borderLeftStyle?: string;
   borderLeftColor?: string;
   borderLeftWidth?: string;
   borderBottomStyle?: string;
   borderBottomColor?: string;
   borderBottomWidth?: string;
   borderRightStyle?: string;
   borderRightColor?: string;
   borderRightWidth?: string;
   borderTopCSS?: string;
   borderLeftCSS?: string;
   borderBottomCSS?: string;
   borderRightCSS?: string;
   halignmentEnabled?: boolean;
   valignmentEnabled?: boolean;
   formatEnabled?: boolean;
   decimalFmts?: string[];
   durationPadZeros?: boolean;
}

export class FontInfo {
   fontFamily: string;
   fontSize: string;
   fontStyle: string;
   fontUnderline: string;
   fontStrikethrough: string;
   fontWeight: string;
   smallCaps?: string;
   allCaps?: string;
   subScript?: string;
   supScript?: string;
   shadow?: string;
}

export class AlignmentInfo {
   valign: string;
   halign: string;
   static LEFT = "Left";
   static CENTER = "Center";
   static RIGHT = "Right";
   static TOP = "Top";
   static MIDDLE = "Middle";
   static BOTTOM = "Bottom";
}
