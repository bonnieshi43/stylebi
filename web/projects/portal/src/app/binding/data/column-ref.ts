/*
 * inetsoft-web - StyleBI is a business intelligence web application.
 * Copyright Â© 2024 InetSoft Technology (info@inetsoft.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import { AbstractDataRef } from "../../common/data/abstract-data-ref";
import { DataRefWrapper } from "../../common/data/data-ref-wrapper";
import { DataRef } from "../../common/data/data-ref";
import { AttributeRef } from "../../common/data/attribute-ref";
import { DateRangeRef } from "../../common/data/date-range-ref";
import { Tool } from "../../../../../shared/util/tool";
import { NumericRangeRef } from "../../common/data/numeric-range-ref";
import { GroupRef } from "../../common/data/group-ref";
import { AggregateRef } from "../../common/data/aggregate-ref";

export class ColumnRef extends AbstractDataRef implements DataRefWrapper {
   dataRefModel: DataRef;
   alias: string;
   width: number;
   visible: boolean;
   valid: boolean;
   sql: boolean;
   description: string;
   order?: number;

   public static equal(ref1: ColumnRef, ref2: ColumnRef): boolean {
      if(ref1 === ref2) {
         return true;
      }

      if(ref1 == undefined || ref2 == undefined) {
         return false;
      }

      return ref1.entity === ref2.entity &&
         ref1.attribute === ref2.attribute &&
         ref1.alias === ref2.alias &&
         ref1.width === ref2.width &&
         ref1.visible === ref2.visible &&
         ref1.valid === ref2.valid &&
         ref1.sql === ref2.sql &&
         ref1.dataType === ref2.dataType &&
         ref1.description === ref2.description;
   }

   // compare name only. columnRef many not contain all same attributes when they are
   // created from different point (e.g. join op vs. table info). make sure we can find
   // the column since join is done by table.column instead of full attributes
   public static equalName(ref1: ColumnRef, ref2: ColumnRef): boolean {
      if(ref1 === ref2) {
         return true;
      }

      if(ref1 == undefined || ref2 == undefined) {
         return false;
      }

      return ref1.name == ref2.name;
   }

   /**
    * Get the caption of a column ref.
    *
    * @param ref the ref to get the caption of
    * @returns the ref's caption
    */
   public static getCaption(ref: ColumnRef): string {
      let caption: string;

      if(ref.dataRefModel && (<AttributeRef> ref.dataRefModel).caption) {
         caption = (<AttributeRef> ref.dataRefModel).caption;
      }
      else {
         const entity: string = ref.entity;
         const attr: string = ref.attribute;
         // Bug #23023
         caption = entity == null ? attr : entity.replace(/(OUTER_|_\d+)/g, "") + "." + attr;
      }

      return caption;
   }

   public static getTooltip(_ref: DataRef): string {
      let tooltip: string = "";
      let dataRef: DataRef = ColumnRef.getColumnRef(_ref)?.dataRefModel;
      let att: AttributeRef = this.getAttributeRef(dataRef);

      if(att == null || dataRef == null) {
         return dataRef != null && dataRef.classType == "ExpressionRef" ?
            ColumnRef.getColumnRef(_ref)?.description : tooltip;
      }

      if(dataRef.classType == "NumericRangeRef" || dataRef.classType == "DateRangeRef") {
         tooltip = _ref.view + " (" + att.view + ")" + (!Tool.isEmpty(_ref.description) ?
            "\n" + _ref.description : "");
      }
      else {
         tooltip = "Alias: " + _ref.view + " (" + att.view + ")" + (!Tool.isEmpty(_ref.description) ?
            "\nDescription: " + _ref.description : "");
      }

      return tooltip;
   }

   public static getColumnRef(ref: DataRef): ColumnRef {
      if(ref.classType == "GroupRef") {
         let group: GroupRef = <GroupRef> ref;
         return ColumnRef.getColumnRef(group.ref);
      }
      else if(ref.classType == "AggregateRef") {
         let group: AggregateRef = <AggregateRef> ref;
         return ColumnRef.getColumnRef(group.ref);
      }
      else if(ref.classType == "ColumnRef"){
         return <ColumnRef> ref;
      }

      return null;
   }

   public static getAttributeRef(ref: DataRef): AttributeRef {
      if(!!!ref) {
         return null;
      }

      if(ref.classType == "AttributeRef") {
         return <AttributeRef> ref;
      }
      else if(ref.classType == "DateRangeRef") {
         return <AttributeRef> (<DateRangeRef> ref).ref;
      }
      else if(ref.classType == "NumericRangeRef") {
         return <AttributeRef> (<NumericRangeRef> ref).ref;
      }

      return null;
   }
}
