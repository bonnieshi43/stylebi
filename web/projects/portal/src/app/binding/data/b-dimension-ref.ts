/*
 * This file is part of StyleBI.
 * Copyright (C) 2024  InetSoft Technology
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
import { AbstractBindingRef } from "./abstract-binding-ref";
import { NamedGroupInfo } from "./named-group-info";
import { SortOptionModel } from "./sort-option-model";
import { CalculateInfo } from "./chart/calculate-info";

export class BDimensionRef extends AbstractBindingRef {
   refType?: number;
   dataType: string;
   order: number;
   sortByCol: string;
   rankingOption: string;
   rankingN: string;
   rankingCol: string;
   groupOthers: boolean;
   others: boolean;
   dateLevel: string;
   dateInterval: number;
   timeSeries: boolean;
   namedGroupInfo: NamedGroupInfo;
   customNamedGroupInfo?: NamedGroupInfo;
   summarize: string;
   pageBreak: boolean;
   columnValue: string;
   caption: string;
   manualOrder: string[];
   classType: string;
   sortOptionModel: SortOptionModel;
   specificOrderType?: string;
   rankingAgg?: CalcAggregate;
   sortByColAgg?: CalcAggregate;
}

export interface CalcAggregate {
   baseAggregateName: string;
   calculateInfo: CalculateInfo;
}
