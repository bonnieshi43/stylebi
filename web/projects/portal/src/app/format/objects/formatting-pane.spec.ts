/*
 * This file is part of StyleBI.
 * Copyright (C) 2024  InetSoft Technology
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
import { FormattingPane } from "./formatting-pane.component";
import { FormatInfoModel } from "../../common/data/format-info-model";

const createModel = () => {
    return {
       type: "",
        color: "",
        backgroundColor: "",
        font: null,
        align: null,
        format: "",
    };
};

let pane: FormattingPane;

let model: FormatInfoModel;

describe("FormattingPane Increase Decimal Unit Tests", () => {
    beforeEach(() => {
        pane = new FormattingPane();
        model = createModel();
    });

    it("increase decimal should be enabled for number formats and disabled for the rest", () => {
        pane.formatModel = model;

        model.format = "DateFormat";
        expect(pane.increaseDecimalDisabled()).toBeTruthy();
        model.format = "MessageFormat";
        expect(pane.increaseDecimalDisabled()).toBeTruthy();

        model.format = "CurrencyFormat";
        expect(pane.increaseDecimalDisabled()).toBeFalsy();
        model.format = "PercentFormat";
        expect(pane.increaseDecimalDisabled()).toBeFalsy();
        model.format = "DecimalFormat";
        expect(pane.increaseDecimalDisabled()).toBeFalsy();
    });

   it("non-number format should do nothing", () => {
       pane.formatModel = model;
       model.format = "DateFormat";
       model.formatSpec = null;

       pane.increaseDecimal();

       expect(model.formatSpec).toBeNull();
   });

    it("currency format should increase decimal", () => {
        pane.formatModel = model;
        model.format = "CurrencyFormat";
        model.formatSpec = null;

        pane.increaseDecimal();

        expect(model.formatSpec).toMatchSnapshot();
    });

    it("percent format should increase decimal", () => {
        pane.formatModel = model;
        model.format = "PercentFormat";
        model.formatSpec = null;

        pane.increaseDecimal();

        expect(model.formatSpec).toMatchSnapshot();
    });

   it("simple decimal format should increase decimal", () => {
       pane.formatModel = model;
       model.format = "DecimalFormat";

       const increase = (before, after) => {
           model.formatSpec = before;
           pane.increaseDecimal();
           expect(model.formatSpec).toEqual(after);
       };

       increase("0.0", "0.00");
       increase("0.", "0.0");
       increase("###0.0;(###0.0)", "###0.00;(###0.00)");
       increase("###0%;(###0%)", "###0.0%;(###0.0%)");
       increase("text", ".0text");
   });

   it("empty decimal format should increase decimal", () => {
       pane.formatModel = model;
       model.format = "DecimalFormat";
       model.formatSpec = "";

       pane.increaseDecimal();

       expect(model.formatSpec).toMatchSnapshot();
   });
});

describe("FormattingPane Decrease Decimal Unit Tests", () => {
    beforeEach(() => {
        pane = new FormattingPane();
        model = createModel();
    });

    it("decrease decimal should be enabled for only currency and decimal formats", () => {
        pane.formatModel = model;
        model.formatSpec = "0.0";

        model.format = "DateFormat";
        expect(pane.decreaseDecimalDisabled()).toBeTruthy();
        model.format = "MessageFormat";
        expect(pane.decreaseDecimalDisabled()).toBeTruthy();
        model.format = "PercentFormat";
        expect(pane.decreaseDecimalDisabled()).toBeTruthy();

        model.format = "CurrencyFormat";
        expect(pane.decreaseDecimalDisabled()).toBeFalsy();
        model.format = "DecimalFormat";
        expect(pane.decreaseDecimalDisabled()).toBeFalsy();
    });

    it("decrease decimal should be disabled when the number has no decimal", () => {
        pane.formatModel = model;
        model.format = "DecimalFormat";
        model.formatSpec = "0";

        expect(pane.decreaseDecimalDisabled()).toBeTruthy();
    });

    it("decrease decimal should be enabled when the number has a decimal", () => {
        pane.formatModel = model;
        model.format = "DecimalFormat";

        const enabled = spec => {
            model.formatSpec = spec;
            expect(pane.decreaseDecimalDisabled()).toBeFalsy();
        };

        enabled("0.");
        enabled("0.;(0)");
        enabled("0;(0.)");
        enabled("0.;(0.)");
        enabled("0.0;(0)");
        enabled("0;(0.0)");
        enabled("0.0;(0.0)");
    });

    it("should decrease decimal format", () => {
        pane.formatModel = model;
        model.format = "DecimalFormat";

        const decrease = (before, after) => {
            model.formatSpec = before;
            pane.decreaseDecimal();
            expect(model.formatSpec).toEqual(after);
        };

        decrease("", "");
        decrease("0", "0");
        decrease("0.0", "0");
        decrease("0.", "0");
        decrease("0.;(0)", "0;(0)");
        decrease("0;(0.)", "0;(0)");
        decrease("0.;(0.)", "0;(0)");
        decrease("0.0;(0)", "0;(0)");
        decrease("0;(0.0)", "0;(0)");
        decrease("0.0;(0.0)", "0;(0)");
        decrease("0.0%;(0.0%)", "0%;(0%)");
        decrease("0.00;(0.00)", "0.0;(0.0)");
        decrease("0.00%;(0.00%)", "0.0%;(0.0%)");
        decrease("$0.00;($0.00)", "$0.0;($0.0)");
        decrease("0.##%;(0.##%)", "0.#%;(0.#%)");
        decrease("###0.0#;(###0.0#)", "###0.#;(###0.#)");
        decrease("0.0text", "0text");
    });

   it("non-number format should do nothing", () => {
       pane.formatModel = model;
       model.format = "DateFormat";
       model.formatSpec = null;

       pane.decreaseDecimal();

       expect(model.formatSpec).toBeNull();
   });

   it("currency format should decrease decimal", () => {
       pane.formatModel = model;
       model.format = "CurrencyFormat";
       model.formatSpec = null;

       pane.decreaseDecimal();

       expect(model.formatSpec).toMatchSnapshot();
   });

   it("percent format should do nothing", () => {
       pane.formatModel = model;
       model.format = "PercentFormat";
       model.formatSpec = null;

       pane.decreaseDecimal();

       expect(model.formatSpec).toBeNull();
   });
});
