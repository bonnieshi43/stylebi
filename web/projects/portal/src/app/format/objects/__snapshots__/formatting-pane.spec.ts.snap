// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattingPane Decrease Decimal Unit Tests currency format should decrease decimal 1`] = `"¤#.0"`;

exports[`FormattingPane Increase Decimal Unit Tests currency format should increase decimal 1`] = `"¤#.000"`;

exports[`FormattingPane Increase Decimal Unit Tests empty decimal format should increase decimal 1`] = `"###0.0"`;

exports[`FormattingPane Increase Decimal Unit Tests percent format should increase decimal 1`] = `"#,##0.0%"`;
